### PR TITLE
[dtensor] Fix redistribute(backward_dtype=...) ignoring the backward dtype

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #182032

There was some insane bug that had a very small "fix", but I found it very difficult to understand. The new PR rewrites the internals very explicitly in terms of `op_dtype` and `out_dtype` for each level of backwards, so it is clear what the semantics you are opting into are. And then the fix happens automatically because you are not doing the obviously insane thing.

The history:

* Original addition of forward/backward dtype: https://github.com/pytorch/pytorch/pull/150740
* Bugfix for incorrect behavior with forward dtype: https://github.com/pytorch/pytorch/pull/154975
* Twice differentiable support, which broke backward dtype handling: https://github.com/pytorch/pytorch/pull/160509
* Incomplete bufix: https://github.com/pytorch/pytorch/pull/179495

Authored with Claude.

### CODE REVIEW

File: torch/distributed/tensor/_redistribute.py [hunk 1]
```diff
@@ -1740,8 +1740,8 @@ def redistribute_local_tensor(
 def _redistribute_backward(
     grad_output: "dtensor.DTensor",
     previous_spec: DTensorSpec,
-    original_dtype: torch.dtype | None = None,
-    backward_dtype: torch.dtype | None = None,
+    out_dtype: torch.dtype | None = None,
+    op_dtype: torch.dtype | None = None,
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py]: So I found the `original_dtype` / `forward_dtype` / `backward_dtype` nomenclature confusing. So we're adopting the `spmd_types` convention internally (not for the external API though): https://github.com/meta-pytorch/spmd_types/commit/b3aa1f1cc49b642338d2cc57554cf3f28457539b

File: torch/distributed/tensor/_redistribute.py [hunk 2]
```diff
@@ -1751,25 +1751,26 @@ def _redistribute_backward(
...
+        op_dtype: dtype to run the backward collective at.
         async_op: whether to perform the DTensor redistribute operation
                 asynchronously or not. Default: False
 
     Returns:
         A :class:`torch.Tensor` object.
         A :class:`DTensorSpec` object.
     """
-    if backward_dtype is not None and backward_dtype != grad_output._local_tensor.dtype:
-        local_tensor = grad_output._local_tensor.to(dtype=backward_dtype)
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py:1764]: I think the important observation here is that this function is NOT differentiable (you can see that because it accesses `_local_tensor` directly, which will not be differentiable. This makes it an appropriate spot to do the dtype conversions.

File: torch/distributed/tensor/_redistribute.py [hunk 3]
```diff
@@ -1814,8 +1815,8 @@ def _redistribute_backward(
         async_op=async_op,
     )
 
-    if output.dtype != original_dtype:
-        output = output.to(original_dtype)
+    if output.dtype != out_dtype:
+        output = output.to(out_dtype)
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py:1819]: The changes here net out to be a no op!

File: torch/distributed/tensor/_redistribute.py [hunk 4]
```diff
@@ -1928,18 +1929,25 @@ def forward(  # type: ignore[override]
...
+        ctx.op_dtype = op_dtype
+        out_dtype = out_dtype or ctx.original_dtype
 
         output, spec = _redistribute_backward(
             grad_output,
             previous_spec,
-            ctx.backward_dtype,
-            backward_dtype,
+            out_dtype,
+            op_dtype,
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py]: yes this argument order is correct; I'm going to kwarg'ify

File: torch/distributed/tensor/_api.py [hunk 5]
```diff
@@ -650,12 +650,16 @@ def redistribute(
...
+            forward_dtype (torch.dtype, optional): cast the local tensor to
+                ``forward_dtype`` before running the forward collective.
+                The result DTensor will be in ``forward_dtype``. If ``None``,
+                no conversion is performed. Default: None
+            backward_dtype (torch.dtype, optional): cast the gradient to
+                ``backward_dtype`` before running the backward collective.
+                After the collective the gradient is always cast back to the
+                input DTensor's dtype. If ``None``, the backward collective
+                runs at the input DTensor's dtype (not ``forward_dtype``).
+                Default: None
```
Comment (ezyang) [inline torch/distributed/tensor/_api.py:662]: I attempted to update the docs in https://github.com/pytorch/pytorch/pull/181671/changes but it was wrong because the code was broken when I was reading it. These new docs give us the default semantics we had pre bcce36c (when the bugs were introduced).

TBH, I think the old semantics are also kind of suspicious. But they're not... completely unreasonable. In particular:

* When nothing is set, the default does the intuitive thing (it also does the bfloat16 forwards leads to bfloat16 backwards footgun, c'est la vie--`spmd_types` fixes that)
* When forward is set so that we reduce precision before the the collective, in backwards we increase precision to the input dtype *before* the collective. This was specifically discussed in the original PR:

```
  wanchaol flagged the issue first:

  ▎ if one pass in fwd_dtype only, then forward behavior is right, but in backward we should also convert the grad_out AFTER redistribute
  ▎  back to the original dtensor dtype if it does not match the original dtensor dtype. Otherwise the precision of the DTensor and its
  ▎ grad would mismatch at a later stage

  ruisizhang123 asked if they could just require both or neither:

  ▎ can we enforce an assertion in API register, asking users to specific both fwd_dtype and bwd_dtype, if one of them is not None

  tianyu-l then settled the semantics:

  ▎ I think we should always cast back to the original dtype, even if backward_dtype is given.
  ▎
  ▎ It is debatable, if backward_dtype is not given, should we perform backward redistribute in forward_dtype or original dtype. I
  ▎ personally think it should be the latter.
```

https://github.com/pytorch/pytorch/pull/154975

I'm not gonna touch it, cuz I think `spmd_types` does much better on the choices here :)

File: torch/distributed/tensor/_redistribute.py [hunk 6]
```diff
@@ -1831,30 +1832,65 @@ def _redistribute_backward(
 
 
 class Redistribute(torch.autograd.Function):
+    # Required keys of ``backward_options``. Both control the backward pass:
+    #   op_dtype  -- dtype the backward collective runs at
+    #   out_dtype -- dtype the returned gradient is cast to
+    # Callers must pin concrete dtypes; no inference happens below this layer.
+    _BACKWARD_OPTION_KEYS = frozenset({"op_dtype", "out_dtype"})
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py]: If some one really wants to you can move the defaulting into the function implementation so that we can support arbitrarily many `backward_options`. Mmm... not sure how I feel about this actually.

File: torch/distributed/tensor/_redistribute.py [hunk 7]
```diff
@@ -1831,30 +1832,65 @@ def _redistribute_backward(
...
+
+    # autograd.Function.apply cannot forward kwargs; wrap it so the call sites
+    # spell out op_dtype/out_dtype/backward_options explicitly.
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py]: lol

File: torch/distributed/tensor/_redistribute.py [hunk 8]
```diff
@@ -1957,15 +2026,15 @@ def forward(  # type: ignore[override]
     @staticmethod
     def backward(ctx, grad2_output: "dtensor.DTensor"):  # type: ignore[override]
         previous_spec = ctx.current_spec
-        async_op = ctx.async_op
-        backward_dtype = ctx.backward_dtype or ctx.original_dtype
-
+        # Reuse the same op_dtype one level down; pin out_dtype to the dtype
+        # of the grad we received at this level, since that's what the node
+        # above us expects back.
```
Comment (ezyang) [inline torch/distributed/tensor/_redistribute.py:1990]: This is a policy choice. Another thing we could do is make the user specify it explicitly with nested `backward_options`. This is NOT what the original NestedRedistribute code did, but I'd also argue that the original code was completely broken.

**Comment by ezyang** [comment]
This is no longer AI slop, I have carefully reviewed every line

File: test/distributed/tensor/test_redistribute.py [hunk 9]
```diff
@@ -3067,5 +3091,69 @@ def test_partial_partial_to_shard_shard_uneven(self):
...
+            # which would be a local chunk with no collective).
+            grad_d = DTensor.from_local(
+                torch.ones(64, 8, dtype=out.dtype), mesh, [Partial()]
+            )
+            out.backward(grad_d)
+
+        self.assertEqual(tracer.dtypes_for("all_gather_into_tensor"), [torch.bfloat16])
+        self.assertEqual(tracer.dtypes_for("reduce_scatter_tensor"), [torch.bfloat16])
+        # param.grad should match param's storage dtype.
+        self.assertEqual(local.grad.dtype, torch.float32)
```
Comment (ezyang) [inline test/distributed/tensor/test_redistribute.py:3176]: I didn't review this super carefully but I did check this fails without the fix here.

File: torch/distributed/tensor/_redistribute.py [hunk 10]
```diff
@@ -1881,6 +1896,19 @@ def forward(  # type: ignore[override]
...
+            target_spec = DTensorSpec(
+                device_mesh,
+                target_spec.placements,
+                tensor_meta=TensorMeta(
+                    shape=input.shape,
+                    stride=input.stride(),
+                    dtype=out_dtype,
+                ),
+                use_strided_shard_as_shard_order=target_spec.use_strided_shard_as_shard_order,
+            )
```
Comment (tianyu-l) [inline torch/distributed/tensor/_redistribute.py:1910]: This logic sounds covered by the above two if-else, thus redundant to me. Can we remove?

File: torch/distributed/tensor/_redistribute.py (hunk 8 above)
Comment (tianyu-l) [reply torch/distributed/tensor/_redistribute.py:1990]: I have two general confusion
- why don't we unify everything and use `NestedRedistribute` instead of `Redistribtue`, but only use `NestedRedistribute` for backward.
- why the docstring of `NestedRedistribute` claims "This class is used to make the redistribution of a DTensor twice-differentiable." instead of recursively differentiable. I didn't where the twice come from.

File: test/distributed/tensor/test_redistribute.py [hunk 11]
```diff
@@ -3067,5 +3091,69 @@ def test_partial_partial_to_shard_shard_uneven(self):
...
+        local = torch.randn(16, 8, dtype=torch.float32, requires_grad=True)
+        dt = DTensor.from_local(local, mesh, [Shard(0)])
+
+        tracer = _CollectiveDtypeTracer()
+        with tracer:
+            out = dt.redistribute(
+                mesh,
+                [Replicate()],
+                forward_dtype=torch.bfloat16,
+                backward_dtype=torch.bfloat16,
```
Comment (tianyu-l) [inline test/distributed/tensor/test_redistribute.py:3145]: maybe should at least add a test where backward_dtype is different from forward_dtype

**Review by tianyu-l — APPROVED** [review]
Nice progress to unblock. Please address comments.

File: torch/distributed/tensor/_redistribute.py (hunk 8 above)
Comment (ezyang) [reply torch/distributed/tensor/_redistribute.py:1990]: * I think potentially after this refactor it is possible to get rid of NestedRedistribute. I didn't do it because it is an orthogonal refactoring. I think one potential barrier to doing this is that the semantics for the forward/backward is different than the backward/double backward (because we don't support the API for arbitrary double backwards), so there does have to be something asymmetric in the two cases. But with a more generic `backward_options` setting this might be solvable.
* It is recursively differentiable, people usually say double backwards as shorthand for that lol.

File: torch/distributed/tensor/_redistribute.py (hunk 10 above)
Comment (ezyang) [reply torch/distributed/tensor/_redistribute.py:1910]: Not redundant.  The above if else is the for the `op_dtype` cast, whereas this one is the `out_dtype` cast.

File: torch/distributed/tensor/_redistribute.py (hunk 8 above)
Comment (ezyang) [reply torch/distributed/tensor/_redistribute.py:1990]: I checked, it's not easy, the first backward has custom behavior for propagating partial, whereas the second backward normalizes this.

Comment (ezyang) [reply torch/distributed/tensor/_redistribute.py:1990]: I'll post this on top ig

File: test/distributed/tensor/test_redistribute.py [hunk 12]
```diff
@@ -859,9 +859,12 @@ def test_replicate_to_partial_different_reduce_ops(self):
...
                     device_mesh,
                     [partial_spec],
-                    op_dtype=dt,
-                    out_dtype=dt,
-                    backward_options={"op_dtype": dt, "out_dtype": dt},
+                    False,
+                    {
+                        "op_dtype": dt,
+                        "out_dtype": dt,
+                        "backward_options": {"op_dtype": dt, "out_dtype": dt},
```
Comment (ezyang) [inline test/distributed/tensor/test_redistribute.py:866]: endorsed

**Comment by ezyang** [comment]
## mergedog handoff

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.

claude was not invoked during this run (CI was green from the start; no merge or fix needed).
